### DISCUSSION TITLE
HA memory leak?

### DISCUSSION THREAD

**Discussion by Captain0Kangaroo** [StealthChesnut/HA-FoxESS-Modbus#27] [General]
Since I moved over from the HACS version to the modbus version my HA seems to have a memory leak. I can't 100% confirm this is the cause but I've not installed anything else since or added any other devices.

It could be a total coincidence but I wondered if anyone else has been experiencing memory issues?

My Pi 4b (4gb) runs out after about 24-36 hours.

I could remove the modbus to confirm but don't want to loose my data given the time of year etc 🤣

**Comment by Captain0Kangaroo** [comment]
So I hashed out the modbus in my configuration.ymal and rebooted to see what happens with my memory as it was getting towards the higher end and it seems to be stable again which might mean I'm seeing a memory leak.

It's only been a few hours but as you can see from the HA graph it's flattened out compared to the previous reboot when it constantly climbed.
![image](https://user-images.githubusercontent.com/96144685/173940326-8c31dc46-d7a0-43b2-be48-941cd0990977.png)

**Comment by Captain0Kangaroo** [comment]
I've just have a look at my HA errors and this is popping up frequently. Could this be any reason for concern? @williamjeccles @StealthChesnut @Daibutt 

Logger: homeassistant.components.modbus.modbus
Source: components/modbus/modbus.py:304
Integration: Modbus ([documentation](https://www.home-assistant.io/integrations/modbus), [issues](https://github.com/home-assistant/home-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+modbus%22))
First occurred: 06:43:39 (52 occurrences)
Last logged: 17:29:01

Pymodbus: FoxESSInverterModbus: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)

**Reply by StealthChesnut** [reply]
what do you have your scan_intervals set to, and how many sensors do you have set up?

Is it possible you've got too many sensors being read too frequently, so it's creating a backlog as HA is putting out more modbus requests than it's getting responses, and not cleaning up the open requests?

**Reply by Captain0Kangaroo** [reply]
> what do you have your scan_intervals set to, and how many sensors do you have set up?
> 
> 
> 
> Is it possible you've got too many sensors being read too frequently, so it's creating a backlog as HA is putting out more modbus requests than it's getting responses, and not cleaning up the open requests?

My scan interval is default as per the modbus file (by the looks of it it's set to 1.

In terms of sensors I think again I'm just using the same as the supplied configuration with the addition of Monthly, Yearly and total PV sums.

![image](https://user-images.githubusercontent.com/96144685/174158298-009bb044-39fb-43e2-9400-8926ee211a52.png)

**Reply by StealthChesnut** [reply]
ok - can you try increasing the scan invervals to 5?
If they're set too low/frequent then HA can get overwhelmed - it does complain in the logs that you can see "performance issues" if it's below 5 seconds.

I've not seen issues with it, and I think Will's install has been fine, but the more configs/installs that exist the more info we can draw on. If this fixes it I'll update the defaults to be less aggressive.

**Reply by Captain0Kangaroo** [reply]
I'll give that a try @StealthChesnut and see how things are.

**Reply by Captain0Kangaroo** [reply]
> ok - can you try increasing the scan invervals to 5?
> 

I think that might have worked. Changed the scan intervals that were 1 to 5 and memory usage seems to have stabilised: 
![image](https://user-images.githubusercontent.com/96144685/174233618-9c2140ac-e889-45f9-abf5-17952045d389.png)

Thanks @StealthChesnut

**Reply by StealthChesnut** [reply]
Good progress. Could you report in what it looks like after a couple of days please? Want to make sure it's stable over a longer period!

**Reply by Captain0Kangaroo** [reply]
> Good progress. Could you report in what it looks like after a couple of days please? Want to make sure it's stable over a longer period!

Will do 👍

**Reply by Captain0Kangaroo** [reply]
> Good progress. Could you report in what it looks like after a couple of days please? Want to make sure it's stable over a longer period!

It's been a few days and whilst it helped a lot with the memory issue I think it's not 100% stopped it.

As you can see in the graph it rapidly increased before I changed the scan interval from 1 to 5 and slowed down after it was changed.

I think I'd probably get about 5-7 days before it would crash or I'd have to reboot HA.

![image](https://user-images.githubusercontent.com/96144685/174497774-8b921fa1-4dd7-405d-988f-c896367ab88b.png)

**Reply by williamjeccles** [reply]
Yes my install is working fine with scan set to 1 second but I have 24GB of ram for Home asssistant.

**Reply by canton7** [reply]
See #123
